File: Codeforces/796D.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, k, d;
    cin >> n >> k >> d;
    vector<char> visited(n + 1);
    vector<pair<int, int>> adj[n + 1];
    queue<int> q;
    for (int i = 0; i < k; i++) {
        int p;
        cin >> p;
        q.push(p);
        visited[p] = true;
    }
    for (int i = 1; i < n; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back({v, i});
        adj[v].push_back({u, i});
    }

    set<int> needed;
    while (!q.empty()) {
        int v = q.front();
        q.pop();
        for (auto u : adj[v]) {
            if (!visited[u.first]) {
                visited[u.first] = true;
                q.push(u.first);
                needed.insert(u.second);
            }
        }
    }

    cout << n - 1 - (int)needed.size() << "\n";
    for (int i = 1; i < n; i++) {
        if (needed.find(i) == needed.end()) {
            cout << i << " ";
        }
    }
}

File: Codeforces/796D.py
from collections import deque

n, k, d = map(int, input().split())
p = list(map(int, input().split()))
adj = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)
q = deque()
needed = set()

for i in range(1, n):
    u, v = map(int, input().split())
    adj[u].append((v, i))
    adj[v].append((u, i))

for st in p:
    q.append(st)
    visited[st] = True

while q:
    v = q.popleft()
    for u, i in adj[v]:
        if not visited[u]:
            visited[u] = True
            q.append(u)
            needed.add(i)

print(n - 1 - len(needed))
for i in range(1, n):
    if i not in needed:
        print(i, end=' ')